### DISCUSSION TITLE
Nullable properties in ITenantInfo

### DISCUSSION THREAD

**Discussion by ChristophHornung** [Finbuckle/Finbuckle.MultiTenant#532] [Ideas]
Hello everyone,

first of all thanks for this great lib, it really helped to quickly build a multi tenant app.

I recently update to the 6.6 version (from 6.5) and wondered about the nullable-annotations surrounding ITenantInfo. All properties of ITeantInfo and TenantInfo are now nullable-annotated. 

This results in a few headaches in my nullable-enabled code. First of all every call to e.g. `ITenantInfo.Id` will now be marked nullable when all of my code relied on `Id` being present. This results in a lot of compiler warnings. 

Now I already use my own implementation of `ITenantInfo` so I could replace all usages of `ITenantInfo` in my code with my own class. But in order for it to properly be non-nullable I had to implicitly implement the interface alongside my own non-nullable properties. I.e.:

``` C#
public string Id { get; set; } = null!;
	
string? ITenantInfo.Id
{
	get { return this.Id; }
	set { this.Id = value!; }
}
```

This is quite some effort to get rid of all nullability warnings and avoid lots of `!` operators in code.

The way I read the library's code it seems it is always assumed that the `Id` and `Identifier` properties are not `null`. Additionally there seems no need to have a setter on the `ITenantInfo`' s properties in the first place. As long as the used store can handle the concrete implementation there seems no use for it.

So a better approach might be to change `ITenantInfo` into:

```C#
public interface ITenantInfo
{
	string Id { get; }
	string Identifier { get; }
	string? Name { get; }
	string? ConnectionString { get; }
}
```

This would allow anyone to use the interface and rely on at least `Id` and `Identifier` to always be non-null and avoid nullability-warnings and additionally to implement the interface in a concrete class that has non-nullable properties (e.g. for `ConnectionString`).

Or did I miss something that makes this change not viable?

**Comment by AndrewTriesToCode** [comment]
Yes I agree with you in practice. The goal with nullable originally was to preserve existing behavior and not really to "improve" things with the feature. This is definitely an area for the next breaking release I'd like to clarify. Will probably remove Connection string alltogether. As I update samples to .NET 6 I expect to see a lot of the issues you mentioned.

Thank you for your input and I'm glad you find the library useful.

**Reply by mcrio** [reply]
Hi, just to give this one a +1. I vote for the immutable interface without nullable properties. Thanks for the great library.
